Mathlib/CategoryTheory/Limits/Shapes/Pullback/Categorical/Basic.lean

```diff
@@ -0,0 +1,354 @@
+
+-/
+
+universe v₁ v₂ v₃ v₄ u₁ u₂ u₃ u₄
+
+namespace CategoryTheory.Limits
+
+attribute [local simp] CatCommSq.iso_hom_naturality  CatCommSq.iso_inv_naturality
+
+section
+
+variable {A : Type u₁} {B : Type u₂} {C : Type u₃}
+  [Category.{v₁} A] [Category.{v₂} B] [Category.{v₃} C]
+  (F : A ⥤ B) (G : C ⥤ B)
+
+/-- The `CategoricalPullback F G` is the category of triples
+`(a : A, c : C, F a ≅ G c)`.
+Morphisms `(a, c, e) ⟶ (a', c', e')` are pairs of morphisms
+`(f₁ : a ⟶ a', f₂ : c ⟶ c')` compatible with the specified
+isomorphisms. -/
+@[kerodon 032Z]
+structure CategoricalPullback where
+  /-- the left element -/
+  left : A
+  /-- the right element -/
+  right : C
+  /-- the structural isomorphism `F.obj left ≅ G.obj right` -/
+  iso : F.obj left ≅ G.obj right
+
+namespace CategoricalPullback
+
+/-- A notation for the categorical pullback. -/
+scoped notation:max L:max " ⊡ " R:max => CategoricalPullback L R
+
+variable {F G}
+
+/-- The Hom types for the categorical pullback are given by pairs of maps compatible with the
+structural isomorphisms. -/
+structure Hom (x y : F ⊡ G) where
+  /-- the left component of `f : Hom x y` is a morphism `x.left ⟶ y.left` -/
+  left : x.left ⟶ y.left
+  /-- the right component of `f : Hom x y` is a morphism `x.right ⟶ y.right` -/
+  right : x.right ⟶ y.right
+  /-- the compatibility condition on `left` and `right` with respect to the structure
+  isompophisms -/
+  w : F.map left ≫ y.iso.hom = x.iso.hom ≫ G.map right := by aesop_cat
+
+attribute [reassoc (attr := simp)] Hom.w
+
+@[simps! id_left id_right comp_left comp_right]
+instance : Category (CategoricalPullback F G) where
+  Hom x y := CategoricalPullback.Hom x y
+  id x :=
+    { left := 𝟙 x.left
+      right := 𝟙 x.right }
+  comp f g :=
+    { left := f.left ≫ g.left
+      right := f.right ≫ g.right }
+
+attribute [reassoc] comp_left comp_right
+
+/-- Naturality square for morphisms in the inverse direction. -/
+@[reassoc (attr := simp)]
+lemma Hom.w' {x y : F ⊡ G} (f : x ⟶ y) :
+    G.map f.right ≫ y.iso.inv = x.iso.inv ≫ F.map f.left := by
+  rw [Iso.comp_inv_eq, Category.assoc, Eq.comm, Iso.inv_comp_eq, f.w]
+
+attribute [local ext] Hom in
+/-- Extensionnality principle for morphisms in `CategoricalPullback F G`. -/
+@[ext]
+theorem hom_ext {x y : F ⊡ G} {f g : x ⟶ y}
+    (hₗ : f.left = g.left) (hᵣ : f.right = g.right) : f = g := by
+  apply Hom.ext <;> assumption
+
+section
+
+variable (F G)
+
+/-- `CategoricalPullback.πₗ F G` is the left projection `CategoricalPullback F G ⥤ A`. -/
+@[simps]
+def πₗ : F ⊡ G ⥤ A where
+  obj x := x.left
+  map f := f.left
+
+/-- `CategoricalPullback.πᵣ F G` is the right projection `CategoricalPullback F G ⥤ C`. -/
+@[simps]
+def πᵣ : F ⊡ G ⥤ C where
+  obj x := x.right
+  map f := f.right
+
+/-- The canonical categorical commutative square in which `CategoricalPullback F G` sits. -/
+instance catCommSq : CatCommSq (πₗ F G) (πᵣ F G) F G where
+  iso' := NatIso.ofComponents (fun x ↦ x.iso)
```

Comment:
This PR could be an opportunity to "unprime" `CatCommSq.iso'`?

Reply:
(Hopefully, the `catCommSq_iso_hom_app` lemmas could be autogenerated?)

Reply:
(By the way, I asked [here](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/Depriming.20fields/near/523318206) for a list of primed fields you and other maintainers would like to see unprimed.)